stetho/src/main/java/com/facebook/stetho/common/android/AccessibilityUtil.java

```diff
@@ -0,0 +1,252 @@
+ * {@link com.googlecode.eyesfree.utils.AccessibilityNodeInfoUtils}, but has stripped many features
+ * which are unnecessary here.
+ */
+public final class AccessibilityUtil {
+  private AccessibilityUtil() {
+  }
+
+  /**
+   * Returns whether the specified node has text or a content description.
+   *
+   * @param node The node to check.
+   * @return {@code true} if the node has text.
+   */
+  public static boolean hasText(AccessibilityNodeInfoCompat node) {
+    if (node == null) {
+      return false;
+    }
+
+    return !TextUtils.isEmpty(node.getText()) || !TextUtils.isEmpty(node.getContentDescription());
+  }
+
+  /**
+   * Returns whether the supplied {@link View} and {@link AccessibilityNodeInfoCompat} would
+   * produce spoken feedback if it were accessibility focused.  NOTE: not all speaking nodes are
+   * focusable.
+   *
+   * @param view The {@link View} to evaluate
+   * @param node The {@link AccessibilityNodeInfoCompat} to evaluate
+   * @return {@code true} if it meets the criterion for producing spoken feedback
+   */
+  public static boolean isSpeakingNode(AccessibilityNodeInfoCompat node, View view) {
+    if (node == null || view == null) {
+      return false;
+    }
+
+    if (!node.isVisibleToUser()) {
+      return false;
+    }
+
+    int important = ViewCompat.getImportantForAccessibility(view);
+    if (important == ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO_HIDE_DESCENDANTS ||
+        (important == ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO &&
+            node.getChildCount() <= 0)) {
+      return false;
+    }
+
+    return hasText(node) || node.isCheckable() || hasNonActionableSpeakingDescendants(node, view);
+  }
+
+  /**
+   * Determines if the supplied {@link View} and {@link AccessibilityNodeInfoCompat} has any
+   * children which are not independently accessibility focusable and also have a spoken
+   * description.
+   * <p>
+   * NOTE: Accessibility services will include these children's descriptions in the closest
+   * focusable ancestor.
+   *
+   * @param view The {@link View} to evaluate
+   * @param node The {@link AccessibilityNodeInfoCompat} to evaluate
+   * @return {@code true} if it has any non-actionable speaking descendants within its subtree
+   */
+  public static boolean hasNonActionableSpeakingDescendants(
+      AccessibilityNodeInfoCompat node, View view) {
+
+    if (node == null || view == null || !(view instanceof ViewGroup)) {
+      return false;
+    }
+
+    ViewGroup viewGroup = (ViewGroup) view;
+    for (int i = 0; i < viewGroup.getChildCount(); i++) {
+      View childView = viewGroup.getChildAt(i);
+
+      if (childView == null) {
+        continue;
+      }
+
+      AccessibilityNodeInfoCompat childNode = AccessibilityNodeInfoCompat.obtain();
+      ViewCompat.onInitializeAccessibilityNodeInfo(childView, childNode);
+
+      if (isAccessibilityFocusable(childNode, childView)) {
+        childNode.recycle();
```

Comment:
you can avoid the repetetive calls to `childNode.recycle()` by using try/finally, e.g.

```
AccessibilityNodeInfoCompat childNode = AccessibilityNodeInfoCompat.obtain();

try {
  ViewCompat.onInitializeAccessibilityNodeInfo(childView, childNode);
  if (isAccessibilityFocusable(childNode, childView)) {
    continue;
  }
  if (isSpeakingNode(childNode, childView)) {
    return true;
  }
} finally {
  childNode.recycle();
}
```